File: assignments/assignment-2-solution/src-csharp/TeaShop/Domain/Inventory/InventoryItem.cs
namespace Assignment2Solution.Domain.Inventory;

/// <summary>
///     Represents an item in the tea shop inventory.
/// </summary>
/// <param name="InventoryItemId">The unique identifier for the inventory item.</param>
/// <param name="Name">The name of the tea.</param>
/// <param name="Price">The price per unit.</param>
/// <param name="Quantity">The available quantity.</param>
/// <param name="StarRating">The star rating of the tea.</param>
public record InventoryItem(Guid InventoryItemId, string Name, decimal Price, int Quantity, StarRating StarRating)
{
    /// <summary>
    ///     Gets a value indicating whether the item is in stock.
    /// </summary>
    public bool IsAvailable => Quantity > 0;

    /// <summary>
    ///     Gets the total value of the inventory for this item.
    /// </summary>
    public decimal TotalPrice => Price * Quantity;
}

File: assignments/assignment-2-solution/src-java/src/main/java/assignment2solution/domain/inventory/InventoryItem.java
package assignment2solution.domain.inventory;

import java.math.BigDecimal;
import java.util.Objects;
import java.util.UUID;

/**
 * Represents an item in the tea shop inventory.
 */
public class InventoryItem {
    private final UUID inventoryItemId;
    private final String name;
    private final BigDecimal price;
    private final int quantity;
    private final StarRating starRating;

    /**
     * Initializes a new instance of the {@link InventoryItem} class.
     */
    public InventoryItem(UUID inventoryItemId, String name, BigDecimal price, int quantity, StarRating starRating) {
        this.inventoryItemId = Objects.requireNonNull(inventoryItemId, "inventoryItemId");
        this.name = Objects.requireNonNull(name, "name");
        this.price = Objects.requireNonNull(price, "price");
        this.quantity = quantity;
        this.starRating = Objects.requireNonNull(starRating, "starRating");
    }

    public UUID getInventoryItemId() {
        return inventoryItemId;
    }

    public String getName() {
        return name;
    }

    public BigDecimal getPrice() {
        return price;
    }

    public int getQuantity() {
        return quantity;
    }

    public StarRating getStarRating() {
        return starRating;
    }

    /**
     * Gets a value indicating whether the item is in stock.
     */
    public boolean isAvailable() {
        return quantity > 0;
    }

    /**
     * Gets the total value of the inventory for this item.
     */
    public BigDecimal getTotalPrice() {
        return price.multiply(BigDecimal.valueOf(quantity));
    }
}
